internal/controller/clusterdeployment_controller.go

```diff
@@ -697,21 +697,38 @@ func (r *ClusterDeploymentReconciler) getInfraProvidersNames(ctx context.Context
 	return ips[:len(ips):len(ips)], nil
 }
 
-func (r *ClusterDeploymentReconciler) getCluster(ctx context.Context, namespace, name string, gvk schema.GroupVersionKind) (*metav1.PartialObjectMetadata, error) {
-	opts := &client.ListOptions{
-		LabelSelector: labels.SelectorFromSet(map[string]string{kcm.FluxHelmChartNameKey: name}),
-		Namespace:     namespace,
-	}
-	itemsList := &metav1.PartialObjectMetadataList{}
-	itemsList.SetGroupVersionKind(gvk)
-	if err := r.Client.List(ctx, itemsList, opts); err != nil {
-		return nil, err
-	}
-	if len(itemsList.Items) == 0 {
-		return nil, fmt.Errorf("%s with name %s was not found", gvk.Kind, name)
+func (r *ClusterDeploymentReconciler) getCluster(ctx context.Context, namespace, name string, gvks ...schema.GroupVersionKind) (*metav1.PartialObjectMetadata, error) {
+	var foundClusters []*metav1.PartialObjectMetadata
+
+	for _, gvk := range gvks {
+		opts := &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(map[string]string{kcm.FluxHelmChartNameKey: name}),
+			Namespace:     namespace,
+		}
+		itemsList := &metav1.PartialObjectMetadataList{}
+		itemsList.SetGroupVersionKind(gvk)
+
+		if err := r.Client.List(ctx, itemsList, opts); err != nil {
+			return nil, fmt.Errorf("failed to list %s in namespace %s: %w", gvk.Kind, namespace, err)
+		}
+
+		if len(itemsList.Items) > 1 {
+			return nil, fmt.Errorf("found multiple %s resources with name %s in namespace %s", gvk.Kind, name, namespace)
+		}
+
+		if len(itemsList.Items) == 1 {
+			foundClusters = append(foundClusters, &itemsList.Items[0])
+		}
 	}
 
-	return &itemsList.Items[0], nil
+	switch len(foundClusters) {
+	case 0:
+		return nil, fmt.Errorf("no cluster found with name %s in namespace %s for any of the provided GroupVersionKinds", name, namespace)
+	case 1:
+		return foundClusters[0], nil
+	default:
+		return nil, fmt.Errorf("found multiple clusters with name %s in namespace %s across different GroupVersionKinds", name, namespace)
```

Comment:
I'm not sure we should exit with an error in >1 clusters found. (In theory, there can be >1 cluster declared in one cluster template). Currently, this function is only used to block the removal of cluster deployment if any children object exists. This function should be refactored to either loop through all the GVKs, get all the clusters of each kind, and block the cluster deployment removal if any of the child objects still exist. But it's not in the scope of this task.

For now, let's get the first one found cluster across all GVKs and exit fast with it. And exit with an error if no clusters were found.

Reply:
2 clusters with same name but different ClusterGVK?

Reply:
Yes, it sounds weird, but a template can technically define two clusters with different GVK (ex, AWS+EKS). While our templates don’t, some BYO templates might. I’d prefer to ensure the controller handles this scenario gracefully rather than failing with an error.

Reply:
label selector, hmph, ok

Reply:
something like that?

Reply:
Yes, thank you.